drivers/tee/optee/call.c

```diff
@@ -433,3 +437,125 @@ void optee_disable_shm_cache(struct optee *optee)
 	}
 	optee_cq_wait_final(&optee->call_queue, &w);
 }
+
+size_t optee_correct_params_cnt(size_t num_params)
+{
+	/* If parameters won't fit in on page we need to allocate more
+	 * parameters to fit additinal page addresses.
+	 * We assume that parameters buffer starts on page boundary.
+	 */
+
+	size_t whole_size = OPTEE_MSG_GET_ARG_SIZE(num_params);
+
+	return num_params + (whole_size - 1) / PAGE_SIZE;
+}
+
+/**
+ * optee_fill_mem_ref_param() - fill msg_param structure(s) for given shared
+ * buffer.
+ *
+ * @msg_params - pointer to array of out parameters
+ * @attr - attribute for parameter
+ * @addr - physical address of shared buffer
+ * @size - size of shared buffer
+ *
+ * This function will use OPTEE_MSG_ATTR_FRAGMENT if buffer spans across page
+ * boundary. Caller should ensure that provided array of msg_param have
+ * sufficient size to hold all entries.
+ * This function fills only attributes and addresses. Size and shm ref should
+ * be filled by caller.
+ *
+ * return:
+ *   True - on success
+ *   False - on error
+ */
+bool optee_fill_mem_ref_param(struct optee_msg_param *msg_params, uint32_t attr,
+			      struct page **pages, size_t num_pages, struct tee_shm *shm)
+{
+	size_t i;
+	phys_addr_t pa;
+	struct optee_msg_param *mp = msg_params;
+
+	if (num_pages == 0)
+		return true;
+
+	mp->u.tmem.buf_ptr = page_to_phys(pages[0]);
+	mp->attr = attr | OPTEE_MSG_ATTR_CACHE_PREDEFINED <<
+		OPTEE_MSG_ATTR_CACHE_SHIFT;
+	for (i = 1; i < num_pages; i++) {
+		mp->attr |= OPTEE_MSG_ATTR_FRAGMENT;
+		mp++;
+		/* Check if we filled whole page with addresses */
+		if (((uintptr_t)(mp + 1) & ~PAGE_MASK) == 0) {
```

Comment:
i don't catch this.
should not be `((uintptr_t)mp & PAGE_MASK) != (uintptr_t)(mp + 1) & PAGE_MASK)` ?